semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala

```diff
@@ -0,0 +1,532 @@
+package scala.meta.internal.scalacp
+
+import java.nio.file._
+import scala.collection.mutable
+import scala.meta.internal.metacp._
+import scala.meta.internal.{semanticdb3 => s}
+import scala.meta.internal.semanticdb3.Accessibility.{Tag => a}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Property => p}
+import scala.meta.internal.semanticdb3.SingletonType.{Tag => st}
+import scala.meta.internal.semanticdb3.Type.{Tag => t}
+import scala.reflect.NameTransformer
+import scala.tools.scalap._
+import scala.tools.scalap.scalax.rules.ScalaSigParserError
+import scala.tools.scalap.scalax.rules.scalasig._
+
+object Scalacp {
+  def parse(classfile: ToplevelClassfile): Option[ToplevelInfos] = {
+    // TODO: Parse scalaSig directly from classfile.node
+    // to avoid reading the bytes and parsing the classfile structure twice.
+    val bytes = Files.readAllBytes(classfile.path)
+    val scalapClassfile = ClassFileParser.parse(ByteCode(bytes))
+    ScalaSigParser.parse(scalapClassfile).map { scalaSig =>
+      val toplevels = scalaSig.topLevelClasses ++ scalaSig.topLevelObjects
+      val others = {
+        scalaSig.symbols.toList.flatMap {
+          case sym: SymbolInfoSymbol if !toplevels.contains(sym) => Some(sym)
+          case _ => None
+        }
+      }
+      val stoplevels = toplevels.flatMap(sinfo)
+      val sothers = toplevels.flatMap(spackages).distinct ++ others.flatMap(sinfo)
+      ToplevelInfos(classfile, stoplevels, sothers)
+    }
+  }
+
+  private def spackages(sym: SymbolInfoSymbol): List[s.SymbolInformation] = {
+    val directPackagePath = {
+      val topLevelPath = sym.symbolInfo.owner.path.replace("<empty>", "_empty_")
+      if (topLevelPath.startsWith("_empty_")) topLevelPath
+      else "_root_." + topLevelPath
+    }
+    val transitivePackagePaths = {
+      val directPackageSteps = directPackagePath.split("\\.").toList
+      directPackageSteps
+        .scanLeft("") { (path, step) =>
+          if (path.nonEmpty) path + "." + step
+          else step
+        }
+        .tail
+    }
+    transitivePackagePaths.map { transitivePackagePath =>
+      val (owner, name) = {
+        transitivePackagePath.split("\\.").toList match {
+          case List(name) if name.nonEmpty => ("", name)
+          case ownerSteps :+ name if name.nonEmpty => (ownerSteps.mkString(".") + ".", name)
+          case _ => sys.error(s"unsupported top-level symbol: $sym")
+        }
+      }
+      s.SymbolInformation(
+        symbol = transitivePackagePath + ".",
+        language = Some(s.Language("Scala")),
```

Comment:
I think we should consider making `language`  an enum and add a `version: String` field. I find it awkward to hardcode `"Scala"` and `"Java"` all over the place and it opens up a can of worms related to equality (case-sensitivity etc.)

Reply:
Yep, I was thinking about that yesterday. I've opened a ticket: https://github.com/scalameta/scalameta/issues/1374.